src/sandbox/algorithms/dqn/dqn.py

```diff
@@ -1,42 +1,73 @@
+from __future__ import annotations
+import logging
+from sandbox.action_selection_rules.generic import ActionSelectionRule
+from sandbox.policies.nn_policy import QNetwork
+from sandbox.algorithms.algorithm import Algorithm
 from sandbox.algorithms.dqn.replay_buffer import ReplayMemory, Transition
-from sandbox.algorithms.dqn.policy import QNetwork
+
 
 import gym
 import gym.spaces
-from abc import ABC
-from typing import Callable
+from typing import Callable, TypeVar
 import numpy as np
 import tensorflow as tf
 
+from sandbox.wrappers.discrete_env_wrapper import DiscreteEnvironment
+
 
-class DQNAlgorithm:
+QNetworkType = TypeVar('QNetworkType', bound=QNetwork)
+class DQNAlgorithm(Algorithm[np.ndarray, int, QNetworkType]):
     def __init__(
         self,
-        env: gym.Env,
         memory_size: int,
-        network: type[QNetwork],
-        policy: Callable,
```

Comment:
Zmieniłem nazwę na `action_selection_rule`, ponieważ kminiliśmy z @Agwen12  i doszliśmy do tego, że polityka to jest to co powstaje z algorytmu. 
Natomiast epsilon-greedy, greedy, etc., to są reguły do wyboru akcji przy trenowaniu polityki.

Nie umiem teraz znaleźć dokładnie tej strony na której to jest napisane, ale te screeny co podsyłam też to implukują.


1 - tutaj definują politykę. Ważne jest zwrócić uwagę na to, że polityka jest `with respect to`. Czyli polityka jest czymś co odpowiada za wybór akcji bazując się na czymś
![image](https://user-images.githubusercontent.com/21079319/172080538-6590473b-8110-46e2-9849-66ed9c46f5ac.png)

2- tutaj mówią o performancie `QLearning` i piszą o `epsilon greedy` jak o `action selection`
![image](https://user-images.githubusercontent.com/21079319/172079876-4e97b29c-5b86-4425-9198-06c3d79587f0.png)


Zresztą to też tłumaczy dlaczego najpierw w książce jest o bandytach. 
Rozdział jest potrzebny po to, żeby pokazać jak wybiera się akcje niezależnie od stanu.